### DISCUSSION TITLE
Is possible to publish to different runtimes using --no-build?

### DISCUSSION THREAD

**Discussion by sn4k3** [dotnet/runtime#68008] [Q&A]
I have a script which publish to multiple runtimes: win-x64, linux-x64, arch-x64, rhel-x64, osx-x64, osx-arm64.
All of them execute sequential with `dotnet publish $project -o "$publishFolder/$publishName" -c Release -r $runtime -p:PublishReadyToRun=true --self-contained` but i notice this is a bit slow, it takes +/- 11m to finish. 
My question is if i do `dotnet build $project -c Release` (Generic build without runtime) and then pass `--no-build` on each publish, will it have any negative effect?
Also to speed up publishes, can i parallel the publishes? 1 publish per task/core? Since each publish create it own folder under /bin, it should not be a problem right?

**Comment by huoyaoyuan** [comment]
You can compare the bits of publish results to start.

**Comment by gfoidl** [comment]
In the `csproj` you can specify the runtime-indentifiers (rids), so that at `dotnet restore` (explicitely, or done implicitely by build, etc.) the necessary dependencies are pulled once. Thus it speeds up the binary-generation (at least in my uses).

```xml
<PropertyGroup>
  <RuntimeIdentifiers>win-x64;linux-x64;arch-x64;rhel-x64;osx-x64;osx-arm64</RuntimeIdentifiers>
</PropertyGroup>
```
For the actual publish you need to specify which _rid_ should be used -- as you do already, as can be seen from you question.

> My question is if i do `dotnet build $project -c Release` (Generic build without runtime) and then pass `--no-build` on each publish, will it have any negative effect?

AFAIK yes, there are negative effects, as the `build` will create the "native host" for the app. Publish bundles this together and copies it to the output location.
So you need to do a `build` with the rid specified -- at least implicitely with `dotnet publish` and not specifying `--no-build`.

But try specifying the rids in the project, so that the expensive restore needs to be done only once.

**Reply by sn4k3** [reply]
I tried that, but for some reason it still restore each runtime, maybe because i call each dotnet publish under a wsl bash script... But my restores are cheap, all time is spent here: `UVtools.WPF -> .../UVtools.WPF/bin/Release/net6.0/linux-x64/UVtools.dll`
Before with **net5.0** this was almost instant, now takes forever in this step with **net6.0**
A normal build without publish (dotnet build) takes: `00:00:05.27` in total

I tried to parallel the build script and it did it much faster, it's now about 2m: `Start-Job { bash -c "'build/createRelease.sh' $using:args $using:runtime" }`

**EDIT:** Parallel build does not work well, some packages did not get published at all (missing all content), only 1 survived
